GithubBrowserSample/app/src/main/java/com/android/example/github/repository/CoroutineNetworkBoundResource.kt

```diff
@@ -0,0 +1,152 @@
+import androidx.lifecycle.LiveDataScope
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.Observer
+import androidx.lifecycle.distinctUntilChanged
+import androidx.lifecycle.liveData
+import androidx.lifecycle.map
+import com.android.example.github.api.ApiEmptyResponse
+import com.android.example.github.api.ApiErrorResponse
+import com.android.example.github.api.ApiResponse
+import com.android.example.github.api.ApiSuccessResponse
+import com.android.example.github.vo.Resource
+import com.android.example.github.vo.Status
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.withContext
+import java.util.concurrent.CancellationException
+
+
+fun <ResultType, RequestType> networkBoundResource(
+    saveCallResult: suspend (RequestType) -> Unit,
+    shouldFetch: (ResultType) -> Boolean = { true },
+    loadFromDb: () -> LiveData<ResultType>,
+    fetch: suspend () -> ApiResponse<RequestType>,
+    processResponse: (suspend (ApiSuccessResponse<RequestType>) -> RequestType) = { it.body },
+    onFetchFailed: ((ApiErrorResponse<RequestType>) -> Unit)? = null
+): LiveData<Resource<ResultType>> {
+    return CoroutineNetworkBoundResource(
+        saveCallResult = saveCallResult,
+        shouldFetch = shouldFetch,
+        loadFromDb = loadFromDb,
+        fetch = fetch,
+        processResponse = processResponse,
+        onFetchFailed = onFetchFailed
+    ).asLiveData().distinctUntilChanged() // not super happy about this as the data might be BIG
+}
+
+/**
+ * A [NetworkBoundResource] implementation in corotuines
+ */
+private class CoroutineNetworkBoundResource<ResultType, RequestType>
+@MainThread constructor(
+    private val saveCallResult: suspend (RequestType) -> Unit,
+    private val shouldFetch: (ResultType) -> Boolean = { true },
+    private val loadFromDb: () -> LiveData<ResultType>,
+    private val fetch: suspend () -> ApiResponse<RequestType>,
+    private val processResponse: (suspend (ApiSuccessResponse<RequestType>) -> RequestType),
+    private val onFetchFailed: ((ApiErrorResponse<RequestType>) -> Unit)?
+) {
+    @ExperimentalCoroutinesApi
+    private val result = liveData<Resource<ResultType>> {
+        if (initialValue?.status != Status.SUCCESS) {
+            yield(Resource.loading(initialValue?.data))
+        }
+        val dbSource = loadFromDb()
+        val initialValue = dbSource.await()
+        val willFetch = initialValue == null || shouldFetch(initialValue)
+        if (!willFetch) {
+            // if we won't fetch, just yield existing db values as success
+            yieldSource(dbSource.map {
+                Resource.success(it)
+            })
+        } else {
+            doFetch(dbSource, this)
+        }
+    }
+
+    private suspend fun doFetch(
+        dbSource: LiveData<ResultType>,
+        liveDataScope: LiveDataScope<Resource<ResultType>>
```

Comment:
I'd do it as extension function defined on `liveDataScope` but it is smaller thing

Reply:
if i do that, i lose access to `this` so ends up looking similar. (i would need to use `this@CoroutineNetworkBoundResource` to access the fetch.